### PR TITLE
Add LongCat-Video LoRA training

### PR DESCRIPTION
**Pull request by AkaneTendo25**
Needs more testing, but already functional. 
I made a training test run (T2V) on a 512x512 image dataset (20 items), RTX 3090, 64 gb RAM (Python 3.10, PyTorch 2.6.0, FlashAttn 2.7.4).
Used bf16 checkpoint, dynamic quantization with `--fp8_base`, `--fp8_scaled` flags.
Samples that I got during training confirm model learned style properly (not showing here, cause slightly NSFW).
The speed was approximately 4s/it.

<details><summary>Loss graph</summary>
<p>
image
</p>
</details> 

Here were the exact commands and configs used for training (all launched within activated venv):

<details><summary>Cache latents</summary>
<p>

`python longcat_cache_latents.py --dataset_config _longcat_lora_dataset.toml --vae E:/ComfyUI_windows_portable/ComfyUI/models/vae/wan/wan_2.1_vae.safetensors`

</p>
</details> 

<details><summary>Cache encoder outputs</summary>
<p>

`python longcat_cache_text_encoder_outputs.py --dataset_config _longcat_lora_dataset.toml --text_encoder H:/models/wan21/Wan2.1-T2V-14B/models_t5_umt5-xxl-enc-bf16.pth --batch_size 16 `

</p>
</details> 

<details><summary>Dataset config</summary>
<p>

```
[general]
enable_bucket = true
bucket_no_upscale = false

[[datasets]]
image_directory = "H:/datasets/moebius_test" 
cache_directory = "H:/datasets/moebius_test/cache" 
resolution = [512, 512] 
batch_size = 1 
num_repeats = 1 
caption_extension = ".txt"
```

</p>
</details> 

<details><summary>Sampling prompts</summary>
<p>

```
# prompt 1
Moebius style. Woman with blonde hair is walking on the beach, camera slowly zooms out revealing her naked body.  --w 384 --h 384 --f 45 --d 4 --s 20
```

</p>
</details> 

<details><summary>Training</summary>
<p>

```
accelerate launch --num_cpu_threads_per_process 1 --mixed_precision bf16 longcat_train_network.py ^
    --dit F:/LongCat_full_bf16.safetensors ^
    --vae E:/ComfyUI_windows_portable/ComfyUI/models/vae/wan/wan_2.1_vae.safetensors ^
    --text_encoder H:/models/wan21/Wan2.1-T2V-14B/models_t5_umt5-xxl-enc-bf16.pth ^
    --flash_attn ^
    --fp8_base ^
    --fp8_scaled ^
    --mixed_precision bf16 ^
    --dataset_config G:/samples/longcat-tuner/_longcat_lora_dataset.toml ^
    --gradient_checkpointing ^
    --max_data_loader_n_workers 1 ^
    --persistent_data_loader_workers ^
    --learning_rate 2e-4 ^
    --optimizer_type AdamW8bit ^
    --optimizer_args weight_decay=0.01 ^
    --network_module networks.lora_longcat ^
    --max_grad_norm 1.0 ^
    --network_dim 16 ^
    --network_alpha 16 ^
    --timestep_sampling shift ^
    --discrete_flow_shift 3.0 ^
    --output_dir G:/samples/longcat-tuner/output ^
    --output_name longcat_lora ^
    --log_config ^
    --log_with tensorboard ^
    --logging_dir G:/samples/longcat-tuner/logs ^
    --sample_prompts G:/samples/longcat-tuner/_longcat_lora_sampling.txt ^
    --save_state ^
    --save_every_n_epochs 5 ^
    --max_train_epochs 50 ^
    --sample_every_n_steps 100
```

</p>
</details> 

Probably won't work without Flash Attention yet (i.e. with `--sdpa` flag). I think it is easy to fix <del>, but haven't done it yet.</del> done
Block swapping _should_ work <del>, but I haven't made any extensive tests.</del> confirmed that it works
I2V training was not tested.
Haven't yet tested compatibility of LoRA checkpoints with WanVideoWrapper <del> (but there are no reasons they won't be compatible).</del> keys not compatible, need to convert LoRA

For training I used bf16 checkpoint from [here](https://huggingface.co/AkaneTendo25/LongCat-Video)

<hr />

TODO:

- [x] Remove unused training scripts  
- [x] Add license information
- [x] Ensure training with videos works
- [x] Ensure training does not require Flash Attention
- [ ] Add support for LongCat-Video-Avatar
- [ ] Ensure LoRA compatibility with [WanVideoWrapper](https://github.com/kijai/ComfyUI-WanVideoWrapper)
- [ ] Add standalone inference script and remove remaining unused scripts
- [ ] Add documentation

### CODE REVIEW

**Comment by kohya-ss** [comment]
Thank you, this is great!

Is it correct that the files under `longcat_video` directory are licensed under the LongCat-Video license, i.e. the MIT license?

I also have a few requests.

- It seems there are some unnecessary script files, so it would be nice if you could delete them. (I don't think context_parallel and pipeline are necessary for example.)
- It would be great if you could add a generation script to check the model works.
- It would also be helpful if you could add documentation.

I could do them myself, but it would take quite a bit of time.

**Comment by AkaneTendo25** [comment]
@kohya-ss Thank you, I will try to address these issues as soon as possible.

upd. Yes, both the code and weights are licensed under MIT, I'll add an appropriate notice to indicate this.

**Comment by arledesma** [comment]
@LongCatFan1 If there are no changes to the vendor files then you might be able to add the individual files under the following sections to exclude from ruff formatting and checks, which can make it easier to maintain upstream changes. Definitely confirm this with @kohya-ss.

* https://github.com/kohya-ss/musubi-tuner/blob/54d42c261c17fa2dd5138946ba3aebb2e959ae2e/pyproject.toml#L102-L130
* https://github.com/kohya-ss/musubi-tuner/blob/54d42c261c17fa2dd5138946ba3aebb2e959ae2e/pyproject.toml#L140-L174


I visually compared the additions in `src/musubi_tuner/longcat_video/modules/` against https://github.com/meituan-longcat/LongCat-Video/tree/main/longcat_video/modules and didn't notice too many changes (hooks in longcat_video_dit definitely jumped out as different) so it may be irrelevant.

**Comment by AkaneTendo25** [comment]
@arledesma Thank you for the suggestion. There were also changes in the `attention.py` and `rope_3d.py` scripts due to the removal of context parallelization, the block-sparse attention functionality, and the addition of SDPA support.
But your suggestion makes sense, and I will add it if @kohya-ss has no objections.

**Comment by kohya-ss** [comment]
@arledesma Thank you!

@LongCatFan1 It would be helpful if you could update your pyproject.toml. The current version will probably give us a lot of warnings when we run ruff.

**Comment by AkaneTendo25** [comment]
Closing this PR to avoid polluting musubi-tuner PR list with stalled issues. Unfortunately I don't have the time to continue working on LongCat-Video and will be focusing on LTX-2 instead.

If anyone would be interested in seeing this continued, please open an issue on my fork of musubi-tuner, and I will consider finishing it (and possibly implementing training LoRA for Longcat-Video-Avatar as well).
